chain/src/transaction.rs

```diff
@@ -186,51 +205,144 @@ impl Deserializable for TransactionInput {
+		// sapling format:
+		// overwintered, version (4), version group, inputs, outputs, lock_time, expiry height, joint split
+
+		let version: u32 = reader.read()?;
+		let overwintered = (version & 0x80000000) != 0;
+		let version = (version & 0x7FFFFFFF) as i32;
+
+		let version_group_id = if overwintered {
+			reader.read()?
+		} else {
+			0
+		};
+
+		// reject overwintered transactions of unknown versions
+		let is_overwinter_tx = overwintered && version == OVERWINTER_TX_VERSION && version_group_id == OVERWINTER_TX_VERSION_GROUP_ID;
+		let is_sapling_tx = overwintered && version == SAPLING_TX_VERSION && version_group_id == SAPLING_TX_VERSION_GROUP_ID;
+		if overwintered && !is_overwinter_tx && !is_sapling_tx {
+			return Err(Error::InvalidFormat(format!("Invalid overwinter transaction version: {}, version group: {}",
+				version, version_group_id)));
+		}
+
 		let inputs: Vec<TransactionInput> = reader.read_list()?;
 		let outputs = reader.read_list()?;
 		let lock_time = reader.read()?;
-		let joint_split = if version >= 2 {
-			deserialize_joint_split(reader)?
+
+		let expiry_height = if is_overwinter_tx || is_sapling_tx {
+			reader.read()?
+		} else {
+			0
+		};
+
+		let mut sapling = if is_sapling_tx {
+			let amount = reader.read()?;
+			let spends = reader.read_list()?;
+			let outputs = reader.read_list()?;
+			Some(Sapling {
+				amount,
+				spends,
+				outputs,
+				..Default::default()
+			})
+		} else {
+			None
+		};
+
+		let join_split = if version >= SPROUT_TX_VERSION {
+			let use_groth = overwintered && version >= SAPLING_TX_VERSION;
+			deserialize_join_split(reader, use_groth)?
 		} else {
 			None
 		};
 
+		if let Some(sapling) = sapling.as_mut() {
+			if !sapling.spends.is_empty() || !sapling.outputs.is_empty() {
+				sapling.binding_sig = reader.read()?;
+			}
+		}
+
 		Ok(Transaction {
-			version: version,
-			inputs: inputs,
-			outputs: outputs,
-			lock_time: lock_time,
-			joint_split: joint_split,
+			overwintered,
+			version,
+			version_group_id,
+			inputs,
+			outputs,
+			lock_time,
+			expiry_height,
+			join_split,
+			sapling,
 		})
 	}
 }
 
 #[cfg(test)]
 mod tests {
+	use hex::ToHex;
 	use hash::H256;
-	use ser::{Serializable};
+	use ser::{Serializable, serialize};
 	use super::Transaction;
 
-	// real transaction from block 80000
+	// real transaction from BTC block 80000
 	// https://blockchain.info/rawtx/5a4ebf66822b0b2d56bd9dc64ece0bc38ee7844a23ff1d7320a88c5fdb2ad3e2
 	// https://blockchain.info/rawtx/5a4ebf66822b0b2d56bd9dc64ece0bc38ee7844a23ff1d7320a88c5fdb2ad3e2?format=hex
 	#[test]
-	fn test_transaction_reader() {
-		let t: Transaction = "0100000001a6b97044d03da79c005b20ea9c0e1a6d9dc12d9f7b91a5911c9030a439eed8f5000000004948304502206e21798a42fae0e854281abd38bacd1aeed3ee3738d9e1446618c4571d1090db022100e2ac980643b0b82c0e88ffdfec6b64e3e6ba35e7ba5fdd7d5d6cc8d25c6b241501ffffffff0100f2052a010000001976a914404371705fa9bd789a2fcd52d2c580b65d35549d88ac00000000".into();
+	fn test_btc_transaction() {
```

Comment:
regular tx serialization and sample from zcash blockchain would be more valid for the context?